flecsi/data/hpx/policy.hh

```diff
@@ -0,0 +1,61 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/futures.hpp>
+
+#include "flecsi/data/local/storage.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/util/types.hh"
+
+#include <unordered_map>
+
+namespace flecsi {
+namespace data {
+
+namespace hpx {
+struct data_storage {
```

Comment:
I looked into this some more. Implementing it the way you suggest would require more code duplication (`local::region_impl` would be backend-specific in this case, possibly more).

Reply:
I don't have an HPX environment with which to test the simple changes to it, but the MPI backend works with this patch on top of 6df66d9e:
```diff
diff --git a/flecsi/data/hpx/policy.hh b/flecsi/data/hpx/policy.hh
index 25b99ef82..40b8f2c22 100644
--- a/flecsi/data/hpx/policy.hh
+++ b/flecsi/data/hpx/policy.hh
@@ -32,30 +32,9 @@
 namespace flecsi {
 namespace data {
 
-namespace hpx {
-struct data_storage {
-  data_storage() = default;
-
-  void init(field_id_t fid) {
-    data[fid];
-    futures[fid];
-  }
-
-  local::detail::storage & get_storage(field_id_t fid) {
-    return data.at(fid);
-  }
-
-  ::hpx::shared_future<void> & get_future(field_id_t fid) {
-    return futures.at(fid);
-  }
-
-private:
-  std::unordered_map<field_id_t, local::detail::storage> data;
-  std::unordered_map<field_id_t, ::hpx::shared_future<void>> futures;
+struct backend_storage : local::detail::storage {
+  ::hpx::shared_future<void> future;
 };
-} // namespace hpx
-
-using backend_storage_data = data::hpx::data_storage;
 
 } // namespace data
 } // namespace flecsi
diff --git a/flecsi/data/local/policy.hh b/flecsi/data/local/policy.hh
index 6c8f0d4ba..e10602767 100644
--- a/flecsi/data/local/policy.hh
+++ b/flecsi/data/local/policy.hh
@@ -49,7 +49,7 @@ struct region_impl {
   // std::vector<>.
   region_impl(size2 s, const fields & fs) : s(std::move(s)), fs(fs) {
     for(auto f : fs) {
-      storages.init(f->fid);
+      storages[f->fid];
     }
   }
 
@@ -57,6 +57,8 @@ struct region_impl {
     return s;
   }
 
+  backend_storage& operator[](field_id_t f) {return storages.at(f);}
+
   // The span is safe because it is used only within a user task while the
   // vectors are resized or destroyed only outside user tasks (though perhaps
   // during execute).
@@ -71,17 +73,13 @@ struct region_impl {
     exec::task_processor_type_t ProcessorType =
       exec::task_processor_type_t::loc>
   util::span<T> get_storage(field_id_t fid, std::size_t nelems) {
-    auto & v = storages.get_storage(fid);
+    auto & v = storages.at(fid);
     std::size_t nbytes = nelems * sizeof(T);
     if(nbytes > v.size())
       v.resize(nbytes);
     return {reinterpret_cast<T *>(v.data<ProcessorType>()), nelems};
   }
 
-  decltype(auto) get_future(field_id_t fid) {
-    return storages.get_future(fid);
-  }
-
   auto get_field_info(field_id_t fid) const {
     for(auto f : fs) {
       if(f->fid == fid)
@@ -95,7 +93,7 @@ private:
   fields fs; // fs[].fid is only unique within a region, i.e. r0.fs[].fid is
              // unrelated to r1.fs[].fid even if they have the same value.
 
-  data::backend_storage_data storages;
+  std::unordered_map<field_id_t,backend_storage> storages;
 };
 
 struct region {
@@ -128,8 +126,4 @@ struct partition {
     return r->get_storage<T, ProcessorType>(fid, nelems);
   }
 
-  decltype(auto) get_future(field_id_t fid) const {
-    return r->get_future(fid);
-  }
-
   template<topo::single_space>
diff --git a/flecsi/data/mpi/policy.hh b/flecsi/data/mpi/policy.hh
index c309ef96f..9c618a6e5 100644
--- a/flecsi/data/mpi/policy.hh
+++ b/flecsi/data/mpi/policy.hh
@@ -24,27 +24,7 @@ namespace data {
 /// Direct data storage.
 /// \ingroup data
 /// \{
-namespace mpi {
-struct data_storage {
-  data_storage() = default;
-
-  void init(size_t i) {
-    data[i];
-  }
-
-  local::detail::storage & get_storage(size_t i) {
-    return data.at(i);
-  }
-
-  // this is never called for the MPI backend
-  void get_future(size_t i);
-
-private:
-  std::unordered_map<field_id_t, local::detail::storage> data;
-};
-} // namespace mpi
-
-using backend_storage_data = mpi::data_storage;
+using backend_storage = local::detail::storage;
 
 /// \}
 } // namespace data
diff --git a/flecsi/exec/hpx/task_prologue.hh b/flecsi/exec/hpx/task_prologue.hh
index 55d7e3e16..9a8301e0d 100644
--- a/flecsi/exec/hpx/task_prologue.hh
+++ b/flecsi/exec/hpx/task_prologue.hh
@@ -90,7 +90,7 @@ protected:
 
     // if the current argument as a valid future associated with it, then
     // it has to be used a dependency for executing the current task
-    auto & field_future = r.get_future(f);
+    auto & field_future = r[f].future;
     if(field_future.valid()) {
       std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
       dependencies.push_back(field_future);
```
It seems like it would be nice to do entirely without the extremely small `data/*/policy.hh` files that result, although it's not much nicer to put actual backend-specific code into `data/backend.hh`.